### PR TITLE
[Features][OMEVV][Modules] - omevv_vcenter_info and omevv_firmware_repository_profile_info

### PR DESCRIPTION
**Pull request by singh-lovepreet1**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

File: plugins/modules/omevv_vcenter_info.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: omevv_vcenter_info
+short_description: Retrieve all or specific vCenter information.
+version_added: "9.8.0"
+description:
+  - This module retrieve all or specific vCenter information.
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: This module allows you to retrieve all or specific vCenter information from (...?)

File: plugins/modules/omevv_vcenter_info.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: omevv_vcenter_info
+short_description: Retrieve all or specific vCenter information.
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: Retrieve all or specific OpenManage Enterprise vCenter information.

File: plugins/modules/omevv_vcenter_info.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+version_added: "9.8.0"
+description:
+  - This module retrieve all or specific vCenter information.
+extends_documentation_fragment:
+  - dellemc.openmanage.ome_auth_options
+options:
+  vcenter_hostname:
+    description:
+      - vCenter IP address or hostname.
+      - If I(vcenter_hostname) is provided, then module will retrieve only the specified vCenter details.
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: - If I(vcenter_hostname) is provided, the module retrieves only specified vCenter information.

File: plugins/modules/omevv_vcenter_info.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+author:
+  - "Lovepreet Singh (@singh-lovepreet1)"
+notes:
+  - This module supports C(check_mode).
+  - This module supports IPv4 and IPv6 addresses.
+'''
+
+EXAMPLES = r'''
+---
+- name: Fetch all vCenter information
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: - name: Retrieve all vCenter information from (..?)

File: plugins/modules/omevv_vcenter_info.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  dellemc.openmanage.omevv_vcenter_info:
+    hostname: "192.168.0.1"
+    username: "username"
+    password: "password"
+    ca_path: "path/to/ca_file"
+
+- name: Fetch specific vCenter information
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: - name: Retrieve specific vCenter information from (...?)

File: plugins/modules/omevv_vcenter_info.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    username: "username"
+    password: "password"
+    ca_path: "path/to/ca_file"
+    vcenter_hostname: xx.xx.xx.xx
+'''
+
+RETURN = r'''
+---
+msg:
+  description: Status of the vCenter information retrieval operation.
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: Status of the vCenter information for the retrieve operation.

File: plugins/modules/omevv_vcenter_info.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  returned: always
+  type: str
+  sample: "Successfully fetched the vCenter information."
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: sample: "Successfully retrieved the vCenter information."

File: plugins/modules/omevv_vcenter_info.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+vcenter_info:
+  description: Information about the vCenter.
```
Comment (reetivel) [inline plugins/modules/omevv_vcenter_info.py]: description: Information on the vCenter.

**Review by reetivel — COMMENTED** [review]
Editorial comments added

File: plugins/modules/omevv_vcenter_info.py (hunk 6 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: Done

File: plugins/modules/omevv_vcenter_info.py (hunk 8 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: DOne

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+#
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+---
+module: omevv_firmware_repository_profile_info
+short_description: Retrieve firmware repository profile information.
```
Comment (reetivel) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: short_description: Retrieve OMEVV firmware repository profile information.

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+version_added: "9.8.0"
+description:
+  - This module allows you to retrieve the firmware repository profile information.
```
Comment (reetivel) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: This module allows you to retrieve the OMEVV firmware repository profile information.

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+extends_documentation_fragment:
+  - dellemc.openmanage.omevv_auth_options
+options:
+  name:
+    description:
+      - Name of the firmware repository profile.
```
Comment (reetivel) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: - Name of the OMEVV firmware repository profile.

**Review by reetivel — COMMENTED** [review]
Editorial comments: Please add "OMEVV" with firmware for complete information.

File: plugins/module_utils/omevv_utils.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+VCENTER_INFO_URI = "/Consoles"
+PROFILE_URI = "/RepositoryProfiles"
+VCENTER_ERROR_MSG = "Unable to fetch the vCenter information."
+PROFILE_ERROR_MSG = "Unable to fetch the firmware respository profile information."
+
+
+class OMEVVINFO:
+    def __init__(self, omevv, module):
+        self.omevv = omevv
+        self.module = module
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]: Are we using this anywhere, the util file should avoid module related things.

File: plugins/module_utils/omevv_utils.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+#
+
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+VCENTER_INFO_URI = "/Consoles"
+PROFILE_URI = "/RepositoryProfiles"
+VCENTER_ERROR_MSG = "Unable to fetch the vCenter information."
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]: Is this Used ? If Not please remove.

File: plugins/module_utils/omevv_utils.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+PROFILE_ERROR_MSG = "Unable to fetch the firmware respository profile information."
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]: Is this Used ? If Not please remove.

File: plugins/module_utils/omevv_utils.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class OMEVVINFO:
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]:
```suggestion
class OMEVVInfo:
```

File: plugins/module_utils/omevv_utils.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+    def get_all_vcenter_info(self):
+        """
+        Retrieves all the vCenter Information.
+
+        """
+        resp = self.omevv.invoke_request('GET', VCENTER_INFO_URI)
+        return resp
+
+    def get_firmware_repository_profile(self):
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]: Good to move this to anew util with the name omevv_firmware_api and add all firmware related API's into it.

Also Enhance this method to get a particular profile based on name, id as required.

File: plugins/module_utils/omevv_utils.py
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py:1]: rename to omevv_info_api

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+    }
+    module = OMEVVAnsibleModule(argument_spec=argument_spec,
+                                supports_check_mode=True)
+    try:
+        with RestOMEVV(module.params) as rest_obj:
+            omevv_obj = OmevvFirmwareProfileInfo(module, rest_obj)
+            resp = omevv_obj.perform_module_operation()
+            if resp['op'] == 'success':
+                module.exit_json(msg=resp['msg'], profile_info=resp['profile_info'])
+            elif resp['op'] == 'skipped':
```
Comment (sachin-apa) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: What is the condition of skip in the info, Info should either return data or empty, there is no need of skip.

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+                module.exit_json(msg=resp['msg'], profile_info=resp['profile_info'], skipped=True)
+            else:
+                module.exit_json(msg=resp['msg'], failed=True)
```
Comment (sachin-apa) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: When would this scenario come?  Why do we have failed true in info module.

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+    except HTTPError as err:
+        error_info = json.load(err)
+        code = error_info.get('errorCode')
+        message = error_info.get('message')
+        if code in ERROR_CODES:
```
Comment (sachin-apa) [inline plugins/modules/omevv_firmware_repository_profile_info.py:205]: What are these error codes ?

File: plugins/module_utils/omevv_utils.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+VCENTER_ERROR_MSG = "Unable to fetch the vCenter information."
+PROFILE_ERROR_MSG = "Unable to fetch the firmware respository profile information."
+
+
+class OMEVVINFO:
+    def __init__(self, omevv, module):
+        self.omevv = omevv
+        self.module = module
+
+    def get_all_vcenter_info(self):
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils.py]: Also Enhance this method to get a particular profile based on name, id as required.

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+ERROR_CODES = ["11008", "12027"]
+
+
+class OmevvFirmwareProfileInfo:
+
+    def __init__(self, module, rest_obj) -> None:
+        self.module = module
+        self.obj = rest_obj
+
+    def search_profile_name(self, profiles_data, profile_name):
```
Comment (sachin-apa) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: Can be moved to util

File: plugins/modules/omevv_firmware_repository_profile_info.py [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+        for profile in profiles_data:
+            if profile.get('profileName') == profile_name:
+                return profile
+        return {}
+
+    def get_profile_info(self, profiles_data, profile_name) -> dict:
```
Comment (sachin-apa) [inline plugins/modules/omevv_firmware_repository_profile_info.py]: Can be moved to util

File: plugins/modules/omevv_vcenter_info.py [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+FAILED_MSG = "Unable to fetch the vCenter information."
+
+
+class OMEVVVCenterInfo:
+
+    def __init__(self, module, rest_obj) -> None:
+        self.module = module
+        self.obj = rest_obj
+
+    def search_vcenter_hostname(self, vcenter_data, vcenter_id):
```
Comment (sachin-apa) [inline plugins/modules/omevv_vcenter_info.py]: Can be moved to util

File: plugins/modules/omevv_vcenter_info.py [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+        for vcenter in vcenter_data:
+            if vcenter.get('consoleAddress') == vcenter_id:
+                return vcenter
+        return {}
+
+    def get_vcenter_info(self, result, vcenter_id) -> dict:
```
Comment (sachin-apa) [inline plugins/modules/omevv_vcenter_info.py]: Can be moved to util

File: plugins/modules/omevv_vcenter_info.py [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+                              supports_check_mode=True)
+    try:
+        with RestOMEVV(module.params) as rest_obj:
+            rest_obj.username = module.params.get("username")
+            rest_obj.password = module.params.get("password")
+            omevv_obj = OMEVVVCenterInfo(module, rest_obj)
+            resp = omevv_obj.perform_module_operation()
+            if resp['op'] == 'success':
+                module.exit_json(msg=resp['msg'], vcenter_info=resp['vcenter_info'])
+            elif resp['op'] == 'skipped':
```
Comment (sachin-apa) [inline plugins/modules/omevv_vcenter_info.py]: Indo module will not skip.

File: plugins/modules/omevv_vcenter_info.py [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+                module.exit_json(msg=resp['msg'], vcenter_info=resp['vcenter_info'], skipped=True)
+            else:
+                module.exit_json(msg=resp['msg'], failed=True)
```
Comment (sachin-apa) [inline plugins/modules/omevv_vcenter_info.py]: Info module will not fail

**Review by sachin-apa — CHANGES_REQUESTED** [review]

**Comment by sachin-apa** [comment]
Add documentation to all the Class and method

File: plugins/module_utils/omevv_utils.py (hunk 16 above)
Comment (ABHISHEK-SINHA10) [reply plugins/module_utils/omevv_utils.py]: Do proper naming of functions, if it fetches all profile, make sure function name is clear, same for individual profile name fetching functions

File: plugins/module_utils/omevv_utils.py (hunk 12 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Updated

File: plugins/module_utils/omevv_utils.py (hunk 14 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Removed

File: plugins/module_utils/omevv_utils.py (hunk 15 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Updated

File: plugins/module_utils/omevv_utils.py (hunk 16 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Function is changed to handle both scenarios.

File: plugins/module_utils/omevv_utils.py (hunk 13 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Removed

File: plugins/modules/omevv_firmware_repository_profile_info.py (hunk 17 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_firmware_repository_profile_info.py]: Removed

File: plugins/module_utils/omevv_utils.py (hunk 20 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py]: Updated

File: plugins/modules/omevv_vcenter_info.py (hunk 23 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: Updated

File: plugins/modules/omevv_vcenter_info.py (hunk 24 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: Moved

File: plugins/modules/omevv_vcenter_info.py (hunk 26 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: Removed

File: plugins/modules/omevv_vcenter_info.py (hunk 25 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_vcenter_info.py]: Removed

File: plugins/module_utils/omevv_utils.py
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils.py:1]: Renamed and moved to omevv_utils

File: plugins/modules/omevv_firmware_repository_profile_info.py (hunk 18 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_firmware_repository_profile_info.py]: Removed

File: plugins/modules/omevv_firmware_repository_profile_info.py (hunk 19 above)
Comment (singh-lovepreet1) [reply plugins/modules/omevv_firmware_repository_profile_info.py:205]: Added description for error codes

**Comment by singh-lovepreet1** [comment]
> Add documentation to all the Class and method

Added

File: plugins/module_utils/omevv_utils/omevv_info_utils.py [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+            dict: The vCenter data that matches the given vcenter_id. If no match is found, an empty dictionary is returned.
+        """
+        vcenter_info_spec = [vcenter for vcenter in vcenter_data if vcenter.get('consoleAddress') == vcenter_id]
+        return vcenter_info_spec
+
+    def get_vcenter_info(self, vcenter_id=None):
+        """
+        Retrieves the vCenter information.
+        Parameters:
+            vcenter_id (str, optional): The hsotname of the vCenter. If provided, retrieves the information for the specified vCenter.
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils/omevv_info_utils.py]:
```suggestion
            vcenter_id (str, optional): The hostname of the vCenter. If provided, retrieves the information for the specified vCenter.
```

File: plugins/module_utils/omevv_utils/omevv_info_utils.py [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+        Returns:
+            list: A list of vCenter information. If `vcenter_id` is provided, the list contains the information for the specified vCenter.
+        """
+        resp = self.omevv_obj.invoke_request('GET', VCENTER_INFO_URI)
+        vcenter_info = []
+        if resp.success:
+            vcenter_info = resp.json_data
+            if vcenter_id or vcenter_id == "":
```
Comment (sachin-apa) [inline plugins/module_utils/omevv_utils/omevv_info_utils.py]:
```suggestion
            if vcenter_id or vcenter_id != "":
```

Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils/omevv_info_utils.py]: Addressed

File: plugins/module_utils/omevv_utils/omevv_info_utils.py (hunk 27 above)
Comment (singh-lovepreet1) [reply plugins/module_utils/omevv_utils/omevv_info_utils.py]: Addressed

**Review by sachin-apa — DISMISSED** [review]

**Review by ShivamSh3 — DISMISSED** [review]

**Review by Saksham-Nautiyal — APPROVED** [review]
validation completed

**Review by sachin-apa — APPROVED** [review]

**Review by rajshekarp87 — APPROVED** [review]

**Pull request merged**